package/MDAnalysis/lib/pkdtree.py

```diff
@@ -28,199 +28,192 @@
+            self.coords = coords
+            self.ckdt = cKDTree(self.coords, self.leafsize)
+        self._built = True
+
+    def set_cutoff(self, cutoff):
+        """Change the cutoff to rebuild the tree"""
+        if not self._built:
+            raise RuntimeError('Unbuilt tree. Run tree.set_coords(...)')
+        if self.periodic:
+            if self.cutoff < cutoff:
+                self.cutoff = cutoff
+                self._built = False
+                self.set_coords(self.coords, self.cutoff)
+            else:
+                raise RuntimeError('No need to build')
+        else:
+            raise RuntimeError('No need to build')
 
     def search(self, centers, radius):
-        """Search all points within radius of centers and its periodic images.
-        Wrapping of center coordinates is enforced to enable comparison to
-        wrapped coordinates of points in the tree.
+        """Search all points within radius from centers and their periodic images.
+
+        All the centers coordinates are wrapped around the central cell
+        to enable distance evaluations from points in the tree
+        and their images.
 
         Parameter
         ---------
         centers: array_like (N,3)
-          origins around which to search for neighbors
+          coordinate array to search for neighbors
         radius: float
-          maximum distance around which to search for neighbors. The search
-          radius is half the smallest periodicity if radius exceeds this value
+          maximum distance to search for neighbors.
         """
-        if not self.built:
-            raise RuntimeError('Unbuilt tree. Run tree.set_coords(...) first')
+
+        if not self._built:
+            raise RuntimeError('Unbuilt tree. Run tree.set_coords(...)')
+
         centers = np.asarray(centers)
         if centers.shape == (self.dim, ):
             centers = centers.reshape((1, self.dim))
-        wrapped_centers = apply_PBC(centers, self.box)
+
         self._indices = set()  # clear previous search
-        # possible loop for parallel execution (benchmark before!)
-        for c in itertools.chain(wrapped_centers,
-                                 self.find_images(wrapped_centers, radius)):
-            self.kdt.search_center_radius(c, radius)
-            n = self.kdt.get_count()
-            if n:
-                new_indices = np.empty(n, int)
-                self.kdt.get_indices(new_indices)
-                self._indices.update(new_indices)
-        self._indices = sorted(list(self._indices))
+        if self.periodic:
+            if self.cutoff < radius:
+                raise RuntimeError('Set cutoff greater or equal to the radius.'
+                                   ' Use tree.set_cutoff(...)')
+            # Bring all query points to the central cell
+            wrapped_centers = apply_PBC(centers, self.box)
+            indices = list(self.ckdt.query_ball_point(wrapped_centers,
```

Comment:
Interesting. 

I tried to test the performance of both the cases with different number of query points. Surprisingly, a point query takes longer time when compared with query_ball_tree for a single query. But, for multiple query, the scenario changes for a very small number of queries. I had hoped query_ball_tree would come out victorious, but it seems single point querying is much faster. Since it is not intuitive, please do look at the notebook and suggest if I made a mistake which might be relatively costlier for query_ball_tree as compared to query_ball_point. Here is the link to the notebook.

[https://github.com/ayushsuhane/Benchmarks_Distance/blob/master/Notebooks/Comparison%20of%20Query_ball_tree%20and%20query_ball_point.ipynb](https://github.com/ayushsuhane/Benchmarks_Distance/blob/master/Notebooks/Comparison%20of%20Query_ball_tree%20and%20query_ball_point.ipynb)